### PR TITLE
Cleanup & refactor Dockerfile and init script

### PR DESCRIPTION
**Pull request by sgsunder**
Refactor of Dockerfiles:
- Changed x86 Dockerfile to use latest stable release of Alpine, which matches the ARM Dockerfile
- Don't need to add the edge repos, as both the ssh client and autossh are available on stable from at least v3.9 onwards
- Dropped the `ENV` declaration, they were already being defined in the init script anyway
- Use `apk --no-cache` for a clean one-liner
- No need to to `chmod` the init script if it already is saved with the execute bit in the repo
- Using `CMD` over `ENTRYPOINT`, see https://stackoverflow.com/questions/21553353/
- Renamed 'entrypoint.sh' to 'init' to avoid confusion in regards to the above

Refactor of init script:
- Removed the `chmod` calls to `/id_rsa` and `/known_hosts`. I typically mount these as read-only in my usage, and this causes warnings from these commands to be thrown, but autossh will still function normally, so I don't think they are needed.
- Quickly check to make sure a keyfile is actually mounted
- Add flag `CheckHostIP=no` when `/known_hosts` is used. IMO this makes this a lot more usable for connecting to dynamic IP servers. Readup here: https://unix.stackexchange.com/a/285551
- Avoid repeating code: specify the command string once, and use it for both the log and for actual execution

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -12,19 +12,7 @@ LABEL org.opencontainers.image.ref.name="jnovack/autossh" \
       org.opencontainers.image.revision=$COMMIT \
       org.opencontainers.image.url="https://hub.docker.com/r/jnovack/autossh/"
 
-ENTRYPOINT ["/entrypoint.sh"]
-ADD /entrypoint.sh /entrypoint.sh
-RUN chmod 755 /entrypoint.sh
+RUN apk --no-cache add autossh openssh-client
+COPY /entrypoint.sh /init
```
Comment (jnovack) [inline Dockerfile]: This is not an `init` program. It does not [ensures the usual responsibilities of an init system, such as reaping zombie processes](https://docs.docker.com/engine/reference/run/#specify-an-init-process).

Please revert this change.

File: Dockerfile [hunk 2]
```diff
@@ -12,19 +12,7 @@ LABEL org.opencontainers.image.ref.name="jnovack/autossh" \
       org.opencontainers.image.revision=$COMMIT \
       org.opencontainers.image.url="https://hub.docker.com/r/jnovack/autossh/"
 
-ENTRYPOINT ["/entrypoint.sh"]
```
Comment (jnovack) [inline Dockerfile]: This script IS an [`ENTRYPOINT`](https://docs.docker.com/engine/reference/builder/#entrypoint) and is using the **exec** form intentionally.

There is no other application (that should be) run in this container.  There's no other reason to fire up this container other than this application, thus `ENTRYPOINT` is the proper argument. 

[We are not overwriting any variables from the command-line](https://goinbigdata.com/docker-run-vs-cmd-vs-entrypoint/) and `ENTRYPOINT` provides us a mechanism to run the container AS an executable (using `docker run`).

Please revert this change.

File: Dockerfile [hunk 3]
```diff
@@ -12,19 +12,7 @@ LABEL org.opencontainers.image.ref.name="jnovack/autossh" \
...
-    AUTOSSH_GATETIME=30         \
-    AUTOSSH_POLL=30             \
-    AUTOSSH_FIRST_POLL=30       \
-    AUTOSSH_LOGLEVEL=1
-
-RUN echo "http://dl-cdn.alpinelinux.org/alpine/edge/community/" >> /etc/apk/repositories && \
-    apk update && \
-    apk add --update autossh openssh-client && \
-    rm -rf /var/lib/apt/lists/*
+CMD [ "/init" ]
```
Comment (jnovack) [inline Dockerfile]: Please revert this change.

File: Dockerfile [hunk 4]
```diff
@@ -12,19 +12,7 @@ LABEL org.opencontainers.image.ref.name="jnovack/autossh" \
...
       org.opencontainers.image.url="https://hub.docker.com/r/jnovack/autossh/"
 
-ENTRYPOINT ["/entrypoint.sh"]
-ADD /entrypoint.sh /entrypoint.sh
-RUN chmod 755 /entrypoint.sh
+RUN apk --no-cache add autossh openssh-client
+COPY /entrypoint.sh /init
 
-ENV \
-    TERM=xterm \
```
Comment (jnovack) [inline Dockerfile]: I've had issues in the past removing TERM=xterm with `autossh`.  If that's no longer a problem in newer builds of autossh and ssh, then I am ok with keeping this removed.

File: entrypoint.sh [hunk 5]
```diff
@@ -1,46 +1,52 @@
...
+eval $(ssh-agent -s)
+cat "${SSH_KEY_FILE}" | ssh-add -k -
 
+# Set up known_hosts file if needed
 STRICT_HOSTS_KEY_CHECKING=no
 KNOWN_HOSTS=${SSH_KNOWN_HOSTS:=/known_hosts}
 if [ -f "${KNOWN_HOSTS}" ]; then
-    chmod 0400 ${KNOWN_HOSTS}
-    KNOWN_HOSTS_ARG="-o UserKnownHostsFile=${KNOWN_HOSTS}"
+    KNOWN_HOSTS_ARG="-o UserKnownHostsFile=${KNOWN_HOSTS} -o CheckHostIP=no "
```
Comment (jnovack) [inline entrypoint.sh]: I dislike this as default.  This is forcing everyone who uses to a KNOWN_HOST arg to ignore IP.  A very insecure practice by default.

I would just prefer a new argument at this point.  `AUTOSSH_ARGS`, which is by default blank tacked in the middle of the command somewhere.

File: entrypoint.sh [hunk 6]
```diff
@@ -1,46 +1,52 @@
...
     STRICT_HOSTS_KEY_CHECKING=yes
 fi
 
 # Pick a random port above 32768
 DEFAULT_PORT=$RANDOM
 let "DEFAULT_PORT += 32768"
-echo [INFO] Tunneling ${SSH_HOSTUSER:=root}@${SSH_HOSTNAME:=localhost}:${SSH_TUNNEL_REMOTE:=${DEFAULT_PORT}} to ${SSH_TUNNEL_HOST=localhost}:${SSH_TUNNEL_LOCAL:=22}
```
Comment (jnovack) [inline entrypoint.sh]: Yeah, this was just my newbishness of creating dockerfiles showing when I initially created this years ago.  Thanks for fixing it.

File: entrypoint.sh (hunk 5 above)
Comment (jnovack) [reply entrypoint.sh]: Which naturally should come with a documentation update. **cough cough**

**Comment by jnovack** [comment]
> * Changed x86 Dockerfile to use latest stable release of Alpine, which matches the ARM Dockerfile
> * Don't need to add the edge repos, as both the ssh client and autossh are available on stable from at least v3.9 onwards
> * Use `apk --no-cache` for a clean one-liner
> * No need to to `chmod` the init script if it already is saved with the execute bit in the repo
> * Dropped the `ENV` declaration, they were already being defined in the init script anyway
> * Removed the `chmod` calls to `/id_rsa` and `/known_hosts`.
> * Quickly check to make sure a keyfile is actually mounted
> * Avoid repeating code: specify the command string once, and use it for both the log and for actual execution

👍 


> * Using `CMD` over `ENTRYPOINT`, see https://stackoverflow.com/questions/21553353/
> * Renamed 'entrypoint.sh' to 'init' to avoid confusion in regards to the above

I disagree.  Please see points noted.  In the way we are using this container, it is an `ENTRYPOINT`.  Full revert of these changes please.

> * Add flag `CheckHostIP=no` when `/known_hosts` is used. IMO this makes this a lot more usable for connecting to dynamic IP servers. Readup here: https://unix.stackexchange.com/a/285551

This is a good idea, but should be implemented differently.  In the fashion you are adding it, you are forcing people who use a `KNOWN_HOSTS` file to also `CheckHostIP=no`.  I think a better method of implementation would be to (suggestion) add a new default blank variable called `AUTOSSH_ARGS` and place it in the middle of the command somewhere.  Then, if people want to add `CheckHostIP=no`, they can do so, optionally, through this variable.

File: Dockerfile (hunk 4 above)
Comment (sgsunder) [reply Dockerfile]: I haven't had any issues with this. Must have been a bug that was resolved at some point, I can't imagine why a non-interactive process would need `$TERM` set

File: Dockerfile (hunk 2 above)
Comment (sgsunder) [reply Dockerfile]: I'm holding on this.

> `ENTRYPOINT` provides us a mechanism to run the container AS an executable (using `docker run`).
> There is no other application (that should be) run in this container. There's no other reason to fire up this container other than this application, thus ENTRYPOINT is the proper argument.

Using `CMD` does not prevent or hinder usage of `docker run` or suggest that an container is multi-application. That distinction wouldn't make sense, as the entire container paradigm expects all containers to be single application. Most of Docker's own library images use `CMD`, such as [nginx](https://github.com/nginxinc/docker-nginx/blob/master/stable/alpine/Dockerfile)

`ENTRYPOINT` is meant for containers that wrap and package a single binary, and provide a means for users to pass arguments to that binary without having to re-specify the binary itself. The [StackOverflow link](https://stackoverflow.com/a/21564990) I posted has a good explanation of what the use case of this is. Because the `entrypoint.sh` does not accept any flags or arguments from the command line, there is no reason to use `ENTRYPOINT` here. (And notably, this is why I rename the `entrypoint.sh` to `init`, to avoid this word mangling)

More importantly, it prevents the following issue:

```
docker run -e SSH_TUNNEL_HOST=192.168.1.1 -e SSH_TUNNEL_LOCAL=8080 -v ./mysshkey:/id_rsa jnovack/autossh
```
Will work with both `CMD` and `ENTRYPOINT` notation, where the specified default `CMD` will mean that the `entrypoint.sh` script is used

However, if the user wants to manually override script to set monitor port:
```
docker run -v ./mysshkey:/id_rsa jnovack/autossh autossh -M 12345 -i /id_rsa -R 8080:localhost:80
```
Then this will not work in `ENTRYPOINT` as Docker will try to pass all the overridden command as arguments to `entrypoint.sh`, whereas the `CMD` would just be ignored and the user's command would be executed

**Comment by sgsunder** [comment]
Made change as requested for `CheckHostIP`

Please see above comment made in regards to `ENTRYPOINT` vs `CMD`

**Comment by jnovack** [comment]
> I'm holding on this.

I like you.  I appreciate a good debate!  I'm not saying my repo is perfect, this was made 3 years ago when a younger, less-experienced me wanted to put something out there for others to use. 
 I'm open to growing this project especially as it's now well over 100k downloads (yay!).

> Most of Docker's own library images use CMD, such as nginx, ENTRYPOINT is meant for containers that wrap and package a single binary.

Yes, most of Docker's library images (not "own", they are package by their respective companies, Docker just makes the "official" to be easier to adopt), are using CMD, because they are not single binaries, they are images based on an OS, with the binary in it.  The examples that use ENTRYPOINT are single binary containers (less so), or custom wrappers (even less than that).

In fact, I feel the usage of CMD by those "easier to adopt" official images is also to make Docker "easier to adopt".  Easier to poke around and see how the OS is laid out.

> Because the entrypoint.sh does not accept any flags or arguments from the command line, there is no reason to use ENTRYPOINT here. 

Then that's the problem we have to fix.  Give entrypoint a method to accept arguments from the command line.  [Hashicorp's Consul](https://github.com/hashicorp/docker-consul/blob/b29407b8716fba0c5781f6c6800bf46ef3700cf7/0.X/Dockerfile) is an example of how I want to follow.

```
ENTRYPOINT ["docker-entrypoint.sh"]
CMD ["agent", "-dev", "-client", "0.0.0.0"]
```
```
#!/usr/bin/dumb-init /bin/sh
...
if [ "${1:0:1}" = '-' ]; then
    set -- consul "$@"
fi
...
exec "$@"
```

Rather than hard-coding logic and `CheckHostIP=no`, wouldn't it be easier to specify that from the command line?  `docker run -d jnovack/autossh -o CheckHostIP=no`?

My counter-proposal is to correct the inadequacies of the current entrypoint.sh:
* Add dumb-init to handle SIGTERM gracefully.
* Provide a method for entrypoint.sh to handle and incorporate arguments (such as CheckHostIP and monitor ports).

I feel that correcting the behavior rather than eliminating it could be a "more correct" approach.  Thoughts?

**Comment by sgsunder** [comment]
Ok, I made the changes you requested and returned it back to `ENTRYPOINT`, now that I understand that you actually intend to pass arguments to the script. Also added `dumb-init` like you mentioned and cleaned out some build args that weren't being used/fixed ones that weren't being read properly.

> Provide a method for entrypoint.sh to handle and incorporate arguments (such as CheckHostIP and monitor ports).

Because this would be more of a major change, I will open a seperate PR for this so that you can merge this one quickly.

File: entrypoint.sh [hunk 7]
```diff
@@ -1,46 +1,55 @@
...
 
+# Set up known_hosts file if needed
 STRICT_HOSTS_KEY_CHECKING=no
 KNOWN_HOSTS=${SSH_KNOWN_HOSTS:=/known_hosts}
 if [ -f "${KNOWN_HOSTS}" ]; then
-    chmod 0400 ${KNOWN_HOSTS}
-    KNOWN_HOSTS_ARG="-o UserKnownHostsFile=${KNOWN_HOSTS}"
+    KNOWN_HOSTS_ARG="-o UserKnownHostsFile=${KNOWN_HOSTS} "
+    if [ -z "${SSH_STRICT_HOST_IP_CHECK:=}" ]; then
+        KNOWN_HOSTS_ARG="${KNOWN_HOSTS_ARG}-o CheckHostIP=no "
```
Comment (jnovack) [inline entrypoint.sh]: Please `CheckHostIP=no` from default.

File: entrypoint.sh [hunk 8]
```diff
@@ -1,46 +1,55 @@
...
+"-o ServerAliveCountMax=3 "\
+"-o ExitOnForwardFailure=yes "\
+"-t -t "\
+"${SSH_MODE:=-R} ${SSH_TUNNEL_REMOTE}:${SSH_TUNNEL_HOST}:${SSH_TUNNEL_LOCAL} "\
+"-p ${SSH_HOSTPORT:=22} "\
+"${SSH_HOSTUSER}@${SSH_HOSTNAME}"
+
+# Log to stdout
+INFO_TUNNEL_SRC="${SSH_HOSTUSER:=root}@${SSH_HOSTNAME:=localhost}:${SSH_TUNNEL_REMOTE:=${DEFAULT_PORT}}"
+INFO_TUNNEL_DEST="${SSH_TUNNEL_HOST=localhost}:${SSH_TUNNEL_LOCAL:=22}"
```
Comment (jnovack) [inline entrypoint.sh]: You are defaulting variables (`SSH_HOSTUSER`, `SSH_HOSTNAME`, etc) below, but using them un-defaulted above.

Please move the `INFO` statements above the `COMMAND` statement.

If you run it without variables, you get the following, which is probably not what you want.

```
[INFO] Tunneling root@localhost:44782 to localhost:22
> autossh -M 0 -N -o StrictHostKeyChecking= -o ServerAliveInterval=10 -o ServerAliveCountMax=3
     -o ExitOnForwardFailure=yes -t -t -R :: -p 22 @
```

File: Dockerfile [hunk 9]
```diff
@@ -12,19 +12,10 @@ LABEL org.opencontainers.image.ref.name="jnovack/autossh" \
       org.opencontainers.image.revision=$COMMIT \
```
Comment (jnovack) [inline Dockerfile]: Add one line, this is my fault.  I can do it, but then you'd have to rebase, easier if you do it.

```
    org.opencontainers.image.version=$VERSION \
```

File: hooks/build [hunk 10]
```diff
@@ -7,10 +6,6 @@
 echo "[***] Build hook running"
 
 docker build \
-    --build-arg VERSION=$(git describe --tags --always) \
```
Comment (jnovack) [inline hooks/build]: Per comment above, please leave this in, we don't have versions yet, but I will publish one soon.

**Comment by sgsunder** [comment]
Done!

I also removed `ssh-client` from the `apk add` call. It is a dependency of `autossh` anyway, so there's no need to manually specify it.

What is the purpose of this flag (L37 of `entrypoint.sh`)?:

```
"-t -t "\
```

I can't find any info on it in the manpages of either autossh or the ssh client. Can/should it be removed?

**Comment by jnovack** [comment]
```
-t      Force pseudo-tty allocation.  This can be used to execute arbitrary screen-based
         programs on a remote machine, which can be very useful, e.g. when implementing
         menu services.  Multiple -t options force tty allocation, even if ssh has no local tty.
```

This was necessary in older versions of ssh (we're talking years ago) to prevent the timeout from killing non-tty sessions and helps assist autossh re-establish the connection (because it had trouble figuring out if non-tty sessions were killed).

We need lots of testing, especially long-running tests to make sure removal of this is ok.

File: entrypoint.sh (hunk 7 above)
Comment (jnovack) [reply entrypoint.sh]: Running with no variables causes CheckHostIP to be injected.

```
[INFO] Tunneling root@localhost:50585 to localhost:22
> autossh -M 0 -N -o StrictHostKeyChecking=yes -o UserKnownHostsFile=/known_hosts -o CheckHostIP=no -o ServerAliveInterval=10 -o ServerAliveCountMax=3 -o ExitOnForwardFailure=yes -t -t -R 50585:localhost:22 -p 22 root@localhost
```
